File: app/src/test/java/com/quran/labs/androidquran/presenter/translation/BaseTranslationPresenterTest.java
package com.quran.labs.androidquran.presenter.translation;

import com.quran.labs.androidquran.common.LocalTranslation;
import com.quran.labs.androidquran.common.QuranAyahInfo;
import com.quran.labs.androidquran.common.QuranText;
import com.quran.labs.androidquran.data.VerseRange;
import com.quran.labs.androidquran.presenter.Presenter;

import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static com.google.common.truth.Truth.assertThat;

public class BaseTranslationPresenterTest {
  private BaseTranslationPresenter<TestPresenter> presenter;

  @Before
  public void setupTest() {
    presenter = new BaseTranslationPresenter<>(null, null);
  }

  @Test
  public void testGetTranslationNames() {
    final List<String> databases = Arrays.asList("one.db", "two.db");
    Map<String, LocalTranslation> map = new HashMap<String, LocalTranslation>() {{
      put("one.db", new LocalTranslation(1, "one.db", "One", "First", null, null, 1));
      put("two.db", new LocalTranslation(2, "two.db", "Two", "Second", null, null, 1));
      put("three.db", new LocalTranslation(2, "three.db", "Three", "Third", null, null, 1));
    }};

    String[] translations = presenter.getTranslationNames(databases, map);
    assertThat(translations).hasLength(2);
    assertThat(translations[0]).isEqualTo("First");
    assertThat(translations[1]).isEqualTo("Second");
  }

  @Test
  public void testHashlessGetTranslationNames() {
    final List<String> databases = Arrays.asList("one.db", "two.db");
    final Map<String, LocalTranslation> map = new HashMap<>();

    String[] translations = presenter.getTranslationNames(databases, map);
    assertThat(translations).hasLength(2);
    assertThat(translations[0]).isEqualTo(databases.get(0));
    assertThat(translations[1]).isEqualTo(databases.get(1));
  }

  @Test
  public void testCombineAyahDataOneVerse() throws Exception {
    VerseRange verseRange = new VerseRange(1, 1, 1, 1);
    List<QuranText> arabic = Collections.singletonList(new QuranText(1, 1, "first ayah"));
    List<QuranAyahInfo> info = presenter.combineAyahData(verseRange, arabic,
        Collections.singletonList(Collections.singletonList(new QuranText(1, 1, "translation"))));

    assertThat(info).hasSize(1);
    QuranAyahInfo first = info.get(0);
    assertThat(first.sura).isEqualTo(1);
    assertThat(first.ayah).isEqualTo(1);
    assertThat(first.texts).hasSize(1);
    assertThat(first.arabicText).isEqualTo("first ayah");
    assertThat(first.texts.get(0)).isEqualTo("translation");
  }

  @Test
  public void testCombineAyahDataOneVerseEmpty() throws Exception {
    VerseRange verseRange = new VerseRange(1, 1, 1, 1);
    List<QuranText> arabic = Collections.emptyList();
    List<QuranAyahInfo> info =
        presenter.combineAyahData(verseRange, arabic, Collections.emptyList());
    assertThat(info).hasSize(0);
  }

  @Test
  public void testCombineAyahDataOneVerseNoArabic() throws Exception {
    VerseRange verseRange = new VerseRange(1, 1, 1, 1);
    List<QuranText> arabic = Collections.emptyList();
    List<QuranAyahInfo> info = presenter.combineAyahData(verseRange, arabic,
        Collections.singletonList(Collections.singletonList(new QuranText(1, 1, "translation"))));

    assertThat(info).hasSize(1);
    QuranAyahInfo first = info.get(0);
    assertThat(first.sura).isEqualTo(1);
    assertThat(first.ayah).isEqualTo(1);
    assertThat(first.texts).hasSize(1);
    assertThat(first.arabicText).isNull();
    assertThat(first.texts.get(0)).isEqualTo("translation");
  }

  @Test
  public void testCombineAyahDataArabicEmptyTranslations() throws Exception {
    VerseRange verseRange = new VerseRange(1, 1, 1, 2);
    List<QuranText> arabic = Arrays.asList(
        new QuranText(1, 1, "first ayah"),
        new QuranText(1, 2, "second ayah")
    );
    List<QuranAyahInfo> info = presenter.combineAyahData(verseRange, arabic, new ArrayList<>());
    assertThat(info).hasSize(2);
    assertThat(info.get(0).sura).isEqualTo(1);
    assertThat(info.get(0).ayah).isEqualTo(1);
    assertThat(info.get(0).texts).hasSize(0);
    assertThat(info.get(0).arabicText).isEqualTo("first ayah");
    assertThat(info.get(1).sura).isEqualTo(1);
    assertThat(info.get(1).ayah).isEqualTo(2);
    assertThat(info.get(1).texts).hasSize(0);
    assertThat(info.get(1).arabicText).isEqualTo("second ayah");
  }

  @Test
  public void testEnsureProperTranslations() {
    VerseRange verseRange = new VerseRange(1, 1, 1, 2);

    List<QuranText> text = new ArrayList<>();
    text.add(new QuranText(1, 1, "bismillah"));

    text = presenter.ensureProperTranslations(verseRange, text);
    assertThat(text).hasSize(2);

    QuranText first = text.get(0);
    assertThat(first.sura).isEqualTo(1);
    assertThat(first.ayah).isEqualTo(1);
    assertThat(first.text).isEqualTo("bismillah");

    QuranText second = text.get(1);
    assertThat(second.sura).isEqualTo(1);
    assertThat(second.ayah).isEqualTo(2);
    assertThat(second.text).isEmpty();
  }

  private static class TestPresenter implements Presenter {

    @Override
    public void bind(Object what) {
    }

    @Override
    public void unbind(Object what) {
    }
  }

}

File: app/src/test/java/com/quran/labs/androidquran/presenter/translation/BaseTranslationPresenterTest.kt
package com.quran.labs.androidquran.presenter.translation

import com.google.common.truth.Truth.assertThat
import com.quran.data.core.QuranInfo
import com.quran.data.model.QuranText
import com.quran.data.model.VerseRange
import com.quran.data.pageinfo.common.MadaniDataSource
import com.quran.labs.androidquran.common.TranslationMetadata
import com.quran.labs.androidquran.database.TranslationsDBAdapter
import com.quran.labs.androidquran.model.translation.TranslationModel
import com.quran.labs.androidquran.presenter.Presenter
import com.quran.labs.androidquran.util.TranslationUtil
import com.quran.mobile.translation.model.LocalTranslation
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito

class BaseTranslationPresenterTest {

  private lateinit var presenter: BaseTranslationPresenter<TestPresenter>

  @Before
  fun setupTest() {
    presenter = BaseTranslationPresenter(
        Mockito.mock(TranslationModel::class.java),
        Mockito.mock(TranslationsDBAdapter::class.java),
        object : TranslationUtil(QuranInfo(MadaniDataSource())) {
          override fun parseTranslationText(quranText: QuranText, translationId: Int): TranslationMetadata {
            return TranslationMetadata(quranText.sura, quranText.ayah, quranText.text, translationId)
          }
        },
        QuranInfo(MadaniDataSource())
    )
  }

  @Test
  fun testGetTranslationNames() {
    val databases = listOf("one.db", "two.db")
    val map = object : HashMap<String, LocalTranslation>() {
      init {
        put("one.db", LocalTranslation(1, "one.db", "One", "First", null, "", null, 1, 2))
        put("two.db", LocalTranslation(2, "two.db", "Two", "Second", null, "", null, 1, 2))
        put("three.db", LocalTranslation(3, "three.db", "Three", "Third", null, "", null, 1, 2))
      }
    }

    val translations = presenter.getTranslations(databases, map)
    assertThat(translations).hasLength(2)
    assertThat(translations[0].translator).isEqualTo("First")
    assertThat(translations[1].translator).isEqualTo("Second")
  }

  @Test
  fun testHashlessGetTranslationNames() {
    val databases = listOf("one.db", "two.db")
    val map = HashMap<String, LocalTranslation>()

    val translations = presenter.getTranslations(databases, map)
    assertThat(translations).hasLength(2)
    assertThat(translations[0].filename).isEqualTo(databases[0])
    assertThat(translations[1].filename).isEqualTo(databases[1])
  }

  @Test
  fun testCombineAyahDataOneVerse() {
    val verseRange = VerseRange(1, 1, 1, 1, 1)
    val arabic = listOf(QuranText(1, 1, "first ayah"))
    val info = presenter.combineAyahData(verseRange, arabic,
        listOf(listOf(QuranText(1, 1, "translation"))),
        arrayOf(LocalTranslation(1, "one.db", "One", "First", null, "", null, 1, 2)))

    assertThat(info).hasSize(1)
    val first = info[0]
    assertThat(first.sura).isEqualTo(1)
    assertThat(first.ayah).isEqualTo(1)
    assertThat(first.texts).hasSize(1)
    assertThat(first.arabicText).isEqualTo("first ayah")
    assertThat(first.texts[0].text).isEqualTo("translation")
    assertThat(first.texts[0].localTranslationId).isEqualTo(1)
  }

  @Test
  fun testCombineAyahDataOneVerseEmpty() {
    val verseRange = VerseRange(1, 1, 1, 1, 1)
    val arabic = emptyList<QuranText>()
    val info = presenter.combineAyahData(verseRange, arabic, emptyList(), emptyArray())
    assertThat(info).hasSize(0)
  }

  @Test
  fun testCombineAyahDataOneVerseNoArabic() {
    val verseRange = VerseRange(1, 1, 1, 1, 1)
    val arabic = emptyList<QuranText>()
    val info = presenter.combineAyahData(verseRange, arabic,
        listOf(listOf(QuranText(1, 1, "translation"))), emptyArray())

    assertThat(info).hasSize(1)
    val first = info[0]
    assertThat(first.sura).isEqualTo(1)
    assertThat(first.ayah).isEqualTo(1)
    assertThat(first.texts).hasSize(1)
    assertThat(first.arabicText).isNull()
    assertThat(first.texts[0].text).isEqualTo("translation")
  }

  @Test
  fun testCombineAyahDataArabicEmptyTranslations() {
    val verseRange = VerseRange(1, 1, 1, 2, 2)
    val arabic = listOf(
        QuranText(1, 1, "first ayah"),
        QuranText(1, 2, "second ayah")
    )
    val info = presenter.combineAyahData(verseRange, arabic, ArrayList(), emptyArray())
    assertThat(info).hasSize(2)
    assertThat(info[0].sura).isEqualTo(1)
    assertThat(info[0].ayah).isEqualTo(1)
    assertThat(info[0].texts).hasSize(0)
    assertThat(info[0].arabicText).isEqualTo("first ayah")
    assertThat(info[1].sura).isEqualTo(1)
    assertThat(info[1].ayah).isEqualTo(2)
    assertThat(info[1].texts).hasSize(0)
    assertThat(info[1].arabicText).isEqualTo("second ayah")
  }

  @Test
  fun testEnsureProperTranslations() {
    val verseRange = VerseRange(1, 1, 1, 2, 2)

    val text = listOf(QuranText(1, 1, "bismillah"))
    val result = presenter.ensureProperTranslations(verseRange, text)
    assertThat(result).hasSize(2)

    val first = result[0]
    assertThat(first.sura).isEqualTo(1)
    assertThat(first.ayah).isEqualTo(1)
    assertThat(first.text).isEqualTo("bismillah")

    val second = result[1]
    assertThat(second.sura).isEqualTo(1)
    assertThat(second.ayah).isEqualTo(2)
    assertThat(second.text).isEmpty()
  }

  private class TestPresenter : Presenter<Any> {
    override fun bind(what: Any) {}

    override fun unbind(what: Any) {}
  }
}
